### PR TITLE
Wet bulb temperature integral

### PR DESCRIPTION
**Pull request by gavinevans**
#195 

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/utilities/mathematical_operations.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    def __init__(self, coord_name_to_integrate,
+                 start_point=None, end_point=None,
+                 direction_of_integration="negative"):
+        """
+        Initialise class.
+
+        Args:
+            coord_name_to_integrate (iris.cube.Cube):
+                Name of the coordinate to be integrated.
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: This argument is a string type.

File: lib/improver/utilities/mathematical_operations.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            start_point (float or None):
+                Point at which to start the integration.
+                Default is None.
+            end_point (float or None):
+                Point at which to end the integration.
+                Default is None.
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: start_point and end_point should describe what happens when None is selected; e.g. ".. which starts integration at the first point"

File: lib/improver/utilities/mathematical_operations.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            direction_of_integration (string):
+                Description of the direction in which to integrate.
+                Options are 'positive' or 'negative'.
+                'positive' corresponds to the values within the array
+                increasing as the array index increases.
+                'negative' corresponds to the values within the array
+                decreasing as the array index increases.
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: These three arguments should be listed as "Keyword Args", not "Args" to improve Sphinx docs.

File: lib/improver/utilities/mathematical_operations.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            cube (Iris.cube.Cube):
+                The cube containing the coordinate to check.
+
+        Returns:
+            cube (Iris.cube.Cube):
+                The cube containing a coordinate that is monotonically
+                increasing in the desired direction.
+
+        Raises:
+            ValueError: The chosen coordinate is not monotonic.
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: Is this accompanied by a useful error message?

File: lib/improver/utilities/mathematical_operations.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        lower bound value by 0.5 * stride. The contribution from the
+        uppermost half of the stride and the bottom half of the stride is
+        summed.
+
+        As the coordinate is progressively integrated, the contribution of
+        each stride is cumulatively summed.
+
+        Args:
+            cube (Iris.cube.Cube):
+                The cube containing the coordinate to check.
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: Incorrect? How about "Cube containing data to be integrated"

File: lib/improver/utilities/mathematical_operations.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                    coord_values={self.coord_name_to_integrate:
+                                  lower_bounds})))
+
+        # Determine which cube to copy in order to have the most appropriate
+        # points within the coordinate that is being integrated.
+        if self.direction_of_integration == "positive":
+            integrated_cube = upper_bounds_cube.copy()
+        elif self.direction_of_integration == "negative":
+            integrated_cube = lower_bounds_cube.copy()
+        integrated_cube.data = np.zeros(lower_bounds_cube.shape)
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: Lines 146-171 could be refactored into a function make_working_cubes that defines:
self.upper_bounds_cube
self.lower_bounds_cube
self.integrated_cube

File: lib/improver/utilities/mathematical_operations.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                        continue
+            stride = np.abs(upper_bound - lower_bound)
+            upper_half_of_stride = upper_bounds_slice.data * 0.5 * stride
+            lower_half_of_stride = lower_bounds_slice.data * 0.5 * stride
+            stride_sum += lower_half_of_stride + upper_half_of_stride
+            integrated_slice.data = stride_sum
+            integrated_cubelist.append(integrated_slice.copy())
+
+        # Merge resulting cubes back together
+        integrated_cube = integrated_cubelist.merge_cube()
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: Lines 173-214 could be refactored into a function "perform_integration" that updates self.integrated_cube.

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the convert_distance_into_number_of_grid_cells function from
+ spatial.py."""
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: Replace this with a relevant doc-string

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+from iris.tests import IrisTest
+
+from improver.utilities.mathematical_operations import Integration
+from improver.tests.ensemble_calibration.ensemble_calibration.\
+    helper_functions import set_up_temperature_cube
+
+
+def set_up_height_cube(height_points, cube=set_up_temperature_cube()):
+    """Create a height cube using the existing set_up_temperature_cube
+    utility."""
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: Would this be better: "Create cube with added height dimension.. " ?

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        """Test that the __repr__ returns the expected string."""
+        coord_name = "height"
+        result = str(Integration(coord_name))
+        msg = ('<Integration: coord_name_to_integrate: height, '
+               'start_point: None, end_point: None, '
+               'direction_of_integration: negative>')
+        self.assertEqual(result, msg)
+
+
+class Test_ensure_monotonic_increase_in_chosen_direction(IrisTest):
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: Add a test for the documented "Raises: ValueError"

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        coord_name = "height"
+        direction = "negative"
+        result = (
+            Integration(
+                coord_name, direction_of_integration=direction
+                ).process(self.cube))
+        self.assertIsInstance(result, iris.cube.Cube)
+        self.assertArrayAlmostEqual(
+            result.coord("height").points, np.array([10., 5.]))
+        self.assertArrayAlmostEqual(result.data, expected)
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: Only the last of these three assert statements belongs in this unit test. The first two were performed by test_basic.
This is true for the remaining tests too...

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+
+    """Test the process method."""
+
+    def setUp(self):
+        """Set up the cube."""
+        self.height_points = np.array([5., 10., 20.])
+        cube = set_up_height_cube(self.height_points)[:, 0, :, :, :]
+        data = np.ones(cube.shape)
+        data[1] = np.zeros(cube[1].shape)+2
+        data[2] = np.zeros(cube[2].shape)+3
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: Use np.full(cube[1].shape, fill_value=2)

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+class Test_process(IrisTest):
+
+    """Test the process method."""
+
+    def setUp(self):
+        """Set up the cube."""
+        self.height_points = np.array([5., 10., 20.])
+        cube = set_up_height_cube(self.height_points)[:, 0, :, :, :]
+        data = np.ones(cube.shape)
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: Would be clearer (more consistent) to use data[0] = ...

File: lib/improver/tests/utilities/test_cube_manipulation.py [hunk 14]
```diff
@@ -1297,5 +1301,154 @@ def test_multiple_cell_methods_in_input_cube(self):
...
+class Test_sort_coord_in_cube(IrisTest):
+    """Class to test the sort_coord_in_cube function."""
+
+    def setUp(self):
+        """Set up a cube."""
+        self.ascending_height_points = np.array([5., 10., 20.])
+        cube = set_up_height_cube(self.ascending_height_points)[:, 0, :, :, :]
+        data = np.ones(cube.shape)
+        data[1] = np.zeros(cube[1].shape) + 2
+        data[2] = np.zeros(cube[2].shape) + 3
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_cube_manipulation.py]: See comments in test_mathematical_operations.

File: lib/improver/tests/utilities/test_cube_manipulation.py [hunk 15]
```diff
@@ -1297,5 +1301,154 @@ def test_multiple_cell_methods_in_input_cube(self):
...
+        self.assertArrayAlmostEqual(
+            self.descending_height_points, result.coord(coord_name).points)
+        self.assertDictEqual(
+            result.coord(coord_name).attributes, {"positive": "down"})
+        self.assertArrayAlmostEqual(result.data, expected_data)
+
+    def test_latitude(self):
+        """Test that the sorting successfully sorts the cube based
+        on the points within the given coordinate (latitude).
+        The points in the resulting cube should now be in descending order."""
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_cube_manipulation.py]: What will happen if a longitude coordinate is given that spans the meridian? Should we have a check in the code to make sure there aren't any discontinutities like this? It might well be out-of-scope for IMPROVER.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 16]
```diff
@@ -509,3 +510,82 @@ def process(self, temperature, relative_humidity, pressure):
...
+        wet_bulb_temperature = (
+            WetBulbTemperature(precision=self.precision).process(
+                temperature, relative_humidity, pressure))
+        # Integrate.
+        wet_bulb_temperature_integral = (
+            Integration(
+                self.coord_name_to_integrate, start_point=self.start_point,
+                end_point=self.start_point,
+                direction_of_integration=self.direction_of_integration
+                ).process(wet_bulb_temperature))
```
Comment (MoseleyS) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: I think the initiation of these two child plugins should be in this plugin's __init__ function so that all plugins are initialised at once and all plugins are processed at once.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 17]
```diff
@@ -509,3 +510,82 @@ def process(self, temperature, relative_humidity, pressure):
...
+class WetBulbTemperatureIntegral(object):
+    """Calculate  a wet-bulb temperature integral."""
+
+    def __init__(self, precision=0.005, coord_name_to_integrate="height",
+                 start_point=None, end_point=None,
+                 direction_of_integration="negative"):
+        """
+        Initialise class.
+
+        Args:
```
Comment (MoseleyS) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: These are all Keyword Args.

Comment (gavinevans) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Changed.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 16 above)
Comment (gavinevans) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: The initiation has been moved to the plugin's `__init__` method.

File: lib/improver/tests/utilities/test_cube_manipulation.py (hunk 14 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_cube_manipulation.py]: Changed.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 8 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: Updated.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 9 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: Changed.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 10 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: I've removed the reference to the ValueError from the method, as a ValueError isn't being raised, so this test isn't needed.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 13 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: Changed.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 12 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: Changed.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 11 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: I've removed the test on whether the output is a cube or not, although I've left the check on the coordinate points, as I found it useful for checking that the integration was going in the right direction.

File: lib/improver/utilities/mathematical_operations.py (hunk 1 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Yes, I've edited this.

File: lib/improver/utilities/mathematical_operations.py (hunk 2 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Amended.

File: lib/improver/utilities/mathematical_operations.py (hunk 3 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Changed.

File: lib/improver/utilities/mathematical_operations.py (hunk 4 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: This error isn't actually being raised, so I've removed this from the docstring.

File: lib/improver/utilities/mathematical_operations.py (hunk 5 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Changed.

File: lib/improver/utilities/mathematical_operations.py (hunk 6 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: I'll deal with this in a separate commit.

File: lib/improver/utilities/mathematical_operations.py (hunk 7 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: I'll deal with this in a separate commit.

File: lib/improver/tests/utilities/test_cube_manipulation.py (hunk 15 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_cube_manipulation.py]: An interesting thought. At the moment, I'm decided to capture this by raising a warning if the coordinate is `circular`. However, ideally this is something that we could handle, although it might be sufficiently complicated that it could be handled in a separate PR when the functionality is necessary.

File: lib/improver/utilities/mathematical_operations.py (hunk 6 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Now refactored as suggested.

File: lib/improver/utilities/mathematical_operations.py (hunk 7 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Now refactored as suggested.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 18]
```diff
@@ -521,18 +521,20 @@ def __init__(self, precision=0.005, coord_name_to_integrate="height",
...
                 The precision to which the Newton iterator must converge
                 before returning wet bulb temperatures.
-            coord_name_to_integrate (iris.cube.Cube):
+            coord_name_to_integrate (string):
                 Name of the coordinate to be integrated.
             start_point (float or None):
                 Point at which to start the integration.
-                Default is None.
+                Default is None. If start_point is None, integration is start
+                from the first available point.
```
Comment (MoseleyS) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Grammar could be improved here!

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 19]
```diff
@@ -541,20 +543,18 @@ def __init__(self, precision=0.005, coord_name_to_integrate="height",
...
         """Represent the configured plugin instance as a string."""
-        result = ('<WetBulbTemperatureIntegral: precision: {}, '
-                  'coord_name_to_integrate: {}, start_point: {}, '
-                  'end_point: {}, direction_of_integration: {}>'.format(
-                      self.precision, self.coord_name_to_integrate,
-                      self.start_point, self.end_point,
-                      self.direction_of_integration))
+        result = ('<WetBulbTemperatureIntegral: precision: 0.005, '
+                  'coord_name_to_integrate: height, start_point: None, '
+                  'end_point: None, direction_of_integration: negative>')
```
Comment (MoseleyS) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Surely these values shouldn't be hard-coded?

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 20]
```diff
@@ -168,53 +168,140 @@ def test_descending_coordinate_negative(self):
...
+        expected number of items are returned."""
         coord_name = "height"
         direction = "negative"
         result = (
             Integration(
                 coord_name, direction_of_integration=direction
-                ).process(self.cube))
+                ).prepare_for_integration(self.cube))
+        self.assertIsInstance(result, tuple)
+        self.assertEqual(len(result), 3)
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: I think this unittest should also check that each item in the tuple is an iris cube.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 18 above)
Comment (gavinevans) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Modified the grammar.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 19 above)
Comment (gavinevans) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: True. I've corrected this.

File: lib/improver/utilities/mathematical_operations.py [hunk 21]
```diff
@@ -209,9 +237,44 @@ def process(self, cube):
...
+               upper and lower limits of the integration, as well as as a
+               template cube to put the integrated output.
+            3. Perform the integration using the trapezoidal rule.
+            4. Ensure that the integrated coordinate is a dimension coordinate
+               and ensure that the integrated coordinate is sorted in the
+               desired direction.
+
+        Args:
+            cube (Iris.cube.Cube):
+                Cube containing the data to be integrated.
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: I think we should state here that cube could be modified by this function - the ensure_monotonic function will alter it in-place if necessary.

File: lib/improver/utilities/mathematical_operations.py [hunk 22]
```diff
@@ -47,14 +47,18 @@ def __init__(self, coord_name_to_integrate,
...
-            coord_name_to_integrate (iris.cube.Cube):
+            coord_name_to_integrate (string):
                 Name of the coordinate to be integrated.
+
+        Keyword Args:
             start_point (float or None):
                 Point at which to start the integration.
-                Default is None.
+                Default is None. If start_point is None, integration is start
+                from the first available point.
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: This grammar is wrong too.

File: lib/improver/utilities/mathematical_operations.py [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+            if coord.name() == self.coord_name_to_integrate:
+                integrated_cube = iris.util.new_axis(
+                    integrated_cube, self.coord_name_to_integrate)
+        # Make sure that the order of the coordinate that has been integrated
+        # within the integrated_cube corresponds the direction in which the
+        # cube has been integrated.
+        integrated_cube = (
+            self.ensure_monotonic_increase_in_chosen_direction(
+                integrated_cube))
+        return integrated_cube
```
Comment (MoseleyS) [inline lib/improver/utilities/mathematical_operations.py]: I think the returned cube should set the bounds of the integrated coordinate as well as the points. I don't think this would be difficult to do, although it also needs testing.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 20 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: Added checks for items in the tuple.

File: lib/improver/utilities/mathematical_operations.py (hunk 21 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Added a note.

File: lib/improver/tests/utilities/test_mathematical_operations.py [hunk 24]
```diff
@@ -0,0 +1,466 @@
...
+        integrated. For integration in the positive direction (equivalent to
+        integrating downwards for a height coordinate), the presense of a
+        start_point indicates that the integration may start above the
+        lowest height within the column to be integrated."""
+        expected = np.array(
+            [[[25.00, 25.00, 25.00],
+              [25.00, 25.00, 25.00],
+              [25.00, 25.00, 25.00]]])
+        coord_name = "height"
+        start_point = 8.
```
Comment (carolinejjones) [inline lib/improver/tests/utilities/test_mathematical_operations.py]: If the start point is such that no layers are found then it returns the rather useless error message
 `ValueError: can't merge an empty CubeList`
It would be better if we caught this error. This is also the same for end_point

File: lib/improver/utilities/cube_manipulation.py [hunk 25]
```diff
@@ -614,3 +614,44 @@ def add_renamed_cell_method(cube, orig_cell_method, new_cell_method_name):
...
+               "coordinate span a boundary then the sorting may "
+               "return an undesirable result.".format(coord_to_sort.name()))
+        warnings.warn(msg)
+    dim, = cube.coord_dims(coord_to_sort)
+    index = [slice(None)] * cube.ndim
+    index[dim] = np.argsort(coord_to_sort.points)
+    if order == "descending":
+        index[dim] = index[dim][::-1]
+    if coord in ["height"] and order == "ascending":
+        cube.coord(coord).attributes["positive"] = "up"
```
Comment (carolinejjones) [inline lib/improver/utilities/cube_manipulation.py]: Why is this attribute only added for height. It isn't being used anywhere, do we need it for something. If we do we should add this to the list of added metadata .

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 26]
```diff
@@ -509,3 +510,78 @@ def process(self, temperature, relative_humidity, pressure):
...
+        """
+        # Calculate wet-bulb temperature.
+        wet_bulb_temperature = (
+            self.wet_bulb_temperature_plugin.process(
+                temperature, relative_humidity, pressure))
+        # Integrate.
+        wet_bulb_temperature_integral = (
+            self.integration_plugin.process(wet_bulb_temperature))
+        wet_bulb_temperature_integral.rename("wet_bulb_temperature_integral")
+        wet_bulb_temperature_integral.units = Unit('K m')
```
Comment (carolinejjones) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Surely these units are only correct if we are integrating over height and height is in metres.

File: lib/improver/utilities/mathematical_operations.py [hunk 27]
```diff
@@ -0,0 +1,299 @@
...
+                upper_bounds_slice.coord(
+                    self.coord_name_to_integrate).points.item())
+            lower_bound = (
+                lower_bounds_slice.coord(
+                    self.coord_name_to_integrate).points.item())
+            if not self.start_point and not self.end_point:
+                pass
+            elif self.start_point:
+                if self.direction_of_integration == "positive":
+                    if lower_bound <= self.start_point:
```
Comment (carolinejjones) [inline lib/improver/utilities/mathematical_operations.py]: Should these be < rather than <= . Can you add a few edge case to the unit tests e.g. start_point =5.0 to ensure that this is doing what you think it should be doing.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 26 above)
Comment (gavinevans) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Yes, I've made an edit so that the units are now generated, rather than being hard-coded.

File: lib/improver/tests/utilities/test_mathematical_operations.py (hunk 24 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_mathematical_operations.py]: I'm now catching the exception, in the case that no integration can be done, and raising an error.

File: lib/improver/utilities/cube_manipulation.py (hunk 25 above)
Comment (gavinevans) [reply lib/improver/utilities/cube_manipulation.py]: This attribute is used in Iris for height coordinates, rather being something I've invented for IMPROVER. I think this attribute can be used by Iris, if available, for example there's an mention in http://scitools.org.uk/iris/docs/latest/whatsnew/1.5.html of the "positive" attribute being set to down.  I'm just trying to set it here, in case it's useful metadata later on.

File: lib/improver/utilities/mathematical_operations.py (hunk 27 above)
Comment (gavinevans) [reply lib/improver/utilities/mathematical_operations.py]: Yes, I've changed these to be either < or > and added a couple of unit tests to try to show this.

**Comment by MoseleyS** [comment]
I am happy with these latest changes, although I haven't actually tried running the tests.
